File: projekt3/Pineborscht.py
from Plant import *
from Animal import Animal
from CyberSheep import CyberSheep

class Pineborscht(Plant):
    __PINEBORSCHT_STRENGTH__ = 10
    __COLOR__ = "darkgreen"
    
    def __init__(self, row, col, world):
        super().__init__(row, col, self.__PINEBORSCHT_STRENGTH__,world)
    
    def getName(self):
      return "Pineborscht"
    
    def getColor(self):
        return self.__COLOR__

    def spread(self, newRow, newCol):
        self._world_.setOrganism(newRow, newCol, Pineborscht(newRow , newCol, self._world_))

    def specialAbility(self, other):
        if(other.specialResistance() == True):
            self._world_.removeOrganism(self.getRow(), self.getCol())
            self.die()
            event = other.getName() + " zjada barszcz sosnowskiego"
            self._world_.addEvent(event)
            return True
        else:
            self._world_.removeOrganism(other.getRow(), other.getCol())
            other.die()
        self._world_.removeOrganism(self.getRow(), self.getCol())
        self.die()
        event = other.getName() + "zjada barszcz sosnowskiego i umiera"
        self._world_.addEvent(event)
        return True

    def removeNeighbour(self, row, col):
        if(self._world_.isInsideWorld(row, col) == False):
            return
        neighbour = self._world_.getOrgFromPos(row, col)
        if (neighbour != None and isinstance(neighbour, Animal)== True and neighbour.specialResistance() == False):
            event = neighbour.getName() + " zostaje spalony przez barszcz sosnowskiego"
            self._world_.addEvent(event)
            self._world_.removeOrganism(row, col)
            neighbour.die()

    def takeAction(self):
        self.removeNeighbour(self.getRow() - 1, self.getCol()) #gora
        self.removeNeighbour(self.getRow(), self.getCol() + 1) # prawo
        self.removeNeighbour(self.getRow() + 1, self.getCol()) # dol
        self.removeNeighbour(self.getRow(), self.getCol() - 1) # lewo
        random = randrange(100)
        if (random < 10):
            self.trySpread()

File: projekt2/src/proj/java/model/plant/Pineborscht.java
package proj.java.model.plant;

import java.awt.Color;
import java.util.Random;

import proj.java.model.Organism;
import proj.java.model.Plant;
import proj.java.model.World;

public class Pineborscht extends Plant {
    private static final int PINEBORSCHT_STRENGTH = 10;

    @Override
    protected void spread(int newRow, int newCol, World world) {
        world.setOrganism(newRow, newCol, new Pineborscht());
    }

    @Override
    public int getStrength() {
        return PINEBORSCHT_STRENGTH;
    }

    @Override
    public Color getColor() {
        return Color.BLACK;
    }

    @Override
    public boolean specialAbility(Organism other, World world, boolean specialAblityStopsMovement) {
        world.removeOrganism(other.getRow(), other.getCol());
        world.removeOrganism(this.getRow(), this.getCol());
        String message = String.format("%s zjada barszcz sosnowskiego i umiera", other.getName());
        world.addMessage(message);
        specialAblityStopsMovement = true;
        return true;
    }

    private void removeNeighbour(int row, int col, World world) {
        if(world.isInsideWorld(row, col) == false)
            return;
        if (world.getOrganism(row, col) != null && world.getOrganism(row, col).isAnAnimal()) {
            String message = String.format("%s zostaje spalony przez barszcz sosnowskiego",
                    world.getOrganism(row, col).getName());
            world.addMessage(message);
            world.removeOrganism(row, col);
        }
    }

    @Override
    public void takeAction(World world) {
        removeNeighbour(getRow() - 1, getCol(), world); // gora
        removeNeighbour(getRow(), getCol() + 1, world); // prawo
        removeNeighbour(getRow() + 1, getCol(), world); // dol
        removeNeighbour(getRow(), getCol() - 1, world); // lewo
        var random = new Random();
        if (random.nextInt(100) < 5) {
            trySpread(world);
        }
    }
}
